Add a multi-value input component

bl-nero opened the pull request:
I created this component to handle sets of long strings that make use of `FieldSelectCreate` impractical. The first use of it will be the app [access section of the role editor](https://github.com/gravitational/teleport/pull/47803), which will need to accommodate Azure identity IDs. These strings are very long and can't be effectively dealt with by `FieldSelectCreate`, which is our current go-to solution for editing collections of strings.

The design borrows both from existing label editors, but uses the "X" symbol for removing rows to be consistent with the [new designs](https://www.figma.com/design/v6GunK50D2VC7w7I2FBDNf/Access-(Management)?node-id=5021-40674&node-type=symbol&m=dev). Further updates will also use the same icon for label removal.

Also pending: making the "add more" buttons smaller across the board in both this and label editor components. This is now kept as is to maintain consistency with other UIs so far.

![field-multi-input](https://github.com/user-attachments/assets/cfa8c2b2-3dd8-4ab4-8690-23f766706d7f)

This change depends on https://github.com/gravitational/teleport/pull/47674, but only because it's a part of a bigger PR chain. It's self-contained, so it should be easy to backport.

Followed up by https://github.com/gravitational/teleport/pull/47803.

bl-nero committed:
Add Kubernetes access section to the role editor

bl-nero committed:
Add a multi-value input component

flyinghermit commented on web/packages/shared/components/FieldMultiInput/FieldMultiInput.tsx:
```diff
@@ -0,0 +1,121 @@
+      <Fieldset>
+        {label && <Legend>{label}</Legend>}
+        {value.map((val, i) => (
+          // Note on keys: using index as a key is an anti-pattern in general,
```
>   // Note on keys: using index as a key is an anti-pattern in general,

TIL 👍

avatus commented on web/packages/shared/components/FieldMultiInput/FieldMultiInput.tsx:
```diff
@@ -0,0 +1,121 @@
+  disabled,
+  onChange,
+}: FieldMultiInputProps) {
+  if (value.length === 0) {
+    value = [''];
+  }
```
This doesn't seem right. I don't like updating the prop value here directly, especially with an array of length 1. Because arrays are passed by reference, and this is probably provided by some state, this is technically mutating state directly which we shouldn't do in react. 

Can we think of another way to do this?

bl-nero replied:
It doesn't work this way: if you pass an array by to a function (or set it to a variable), then reassigning to it only changes the local reference, and leaves the original object alone. However, I get it that it may look ugly, so I can just introduce a new variable.

bl-nero replied:
(What would be a violation of rules would be saying `value[0] = ''`.)

avatus commented on web/packages/shared/components/FieldMultiInput/FieldMultiInput.tsx:
```diff
@@ -0,0 +1,121 @@
+              onClick={() => removeItem(i)}
+              disabled={disabled}
+            >
+              <Icon.Cross size="small" color={theme.colors.text.muted} />
```
Love the cross here instead of trashcan, thank you!

avatus commented on web/packages/shared/components/FieldMultiInput/FieldMultiInput.tsx:
```diff
@@ -0,0 +1,121 @@
+          onClick={() => insertItem(value.length)}
+        >
+          <Icon.Plus size="small" mr={2} />
+          Add More
```
Related to my comment about the empty value, perhaps we don't have any input and if length === 0, this says "add field" and then "add more" if greater than 1? Idk, spitballin here.

bl-nero replied:
The way I wrote it, I assumed that there's always at least 1 input element, and if it's empty, it simply means an empty array. Therefore there's no case where `length === 0`. It's an arbitrary decision, I know; it just felt right for the time being, since it's easiest to populate the input (no need to click a button). I know, however, that it leaves an edge case that is unsupported, which is a single empty string. It wasn't really important for the job I was doing at the moment; perhaps if we use this component in more places, we can toggle this behavior with a boolean flag.

avatus replied:
Fair enough

avatus reviewed (commented):
Really sick idea! Left a discussion comment but in general its great

bl-nero committed:
Review

bl-nero committed:
Merge branch 'master' into bl-nero/role-editor-4

bl-nero committed:
Merge branch 'bl-nero/role-editor-4' into bl-nero/multi-input

bl-nero committed:
Merge branch 'master' into bl-nero/role-editor-4

bl-nero committed:
Merge branch 'bl-nero/role-editor-4' into bl-nero/multi-input

bl-nero committed:
Update the k8s operator docs

This also removes the mention of valid values of the Kind field, as I don't
want the external documentation to point to Teleport source files.

bl-nero committed:
Update operator CRDs and Terraform resources

bl-nero committed:
Merge branch 'bl-nero/role-editor-4' into bl-nero/multi-input

bl-nero committed:
Merge branch 'master' into bl-nero/role-editor-4

bl-nero committed:
Merge branch 'bl-nero/role-editor-4' into bl-nero/multi-input

bl-nero committed:
Lint, licenses
